google/cloud/options.h

```diff
@@ -306,6 +307,40 @@ void CheckExpectedOptions(Options const& opts, char const* caller) {
  */
 Options MergeOptions(Options a, Options b);
 
+/**
+ * The prevailing options for the current operation.
+ */
+Options const& CurrentOptions();
+
+/**
+ * RAII object to set/restore the prevailing options for the enclosing scope.
+ *
+ * @code
+ * struct IntOption { using Type = int; };
+ * assert(!internal::CurrentOptions().has<IntOption>());
+ * {
+ *   internal::OptionsSpan span(Options{}.set<IntOption>(1));
+ *   assert(internal::CurrentOptions().get<IntOption>() == 1);
+ *   {
+ *     internal::OptionsSpan span(Options{}.set<IntOption>(2));
+ *     assert(internal::CurrentOptions().get<IntOption>() == 2);
+ *   }
+ *   assert(internal::CurrentOptions().get<IntOption>() == 1);
+ * }
+ * assert(!internal::CurrentOptions().has<IntOption>());
+ * @endcode
+ *
+ * @param opts the `Options` to install.
+ */
+class ABSL_MUST_USE_RESULT OptionsSpan {
+ public:
+  explicit OptionsSpan(Options opts);
+  ~OptionsSpan();
+
+ private:
+  Options opts_;
```

Comment:
nit: If I understand this correctly, `opts_` is actually the previous options that will be _restored_ after the span goes out of scope. That is, `opts_` is _not_ the same as `opts`, which was passed to the c'tor. If you think this distinction is worth capturing, consider renaming this data member to something like `restore_` or `old_opts_` or something.

Reply:
I had considered that, but concluded that the meaning of the member changes with time.  For the majority of its life it is indeed the "options to restore", but after the initializer list is complete it is the "options to install" (i.e., it is the `opts` passed to the constructor), and after restoration it is a "moved from options".  All in all, it seemed better to give it a generic name, and let the code specify its meaning at any point.